### ISSUE TITLE
not thread safe?

### ISSUE THREAD

**Issue by flok99** [libjpeg-turbo/libjpeg-turbo#446]
When running parallel in multiple threads, it crashes.
This has been tested on ubuntu with version 2.0.3-0ubuntu1.20.04.1 of the library, on raspbian (1.5.2-2+b1) and on debian (1.5.2-2+b1)

```
#include <cstring>
#include <stdint.h>
#include <thread>
#include <turbojpeg.h>

constexpr int N = 128;

class example
{
private:
        tjhandle jpegCompressor;

public:
        example() {
                jpegCompressor = tjInitCompress();
        }

        ~example() {
                tjDestroy(jpegCompressor);
        }

        void use() {
                uint8_t pixels[4096], *temp = nullptr;
                long unsigned int len = 0;
                memset(pixels, 0x7f, sizeof pixels);
                tjCompress2(jpegCompressor, pixels, 64, 0, 64, TJPF_RGB, &temp, &len, TJSAMP_444, 1, TJFLAG_FASTDCT);
        }
};

thread_local example e;

int main(int argc, char *argv[])
{
        std::thread *th[N];

        for(;;) {
                for(int i=0; i<N; i++)
                        th[i] = new std::thread([]() { for(int j=0; j<128; j++) e.use(); });

                for(int i=0; i<N; i++) {
                        th[i]->join();
                        delete th[i];
                }
        }

        return 0;
}

```

**Comment by dcommander** [comment]
Change `pixels[4096]` to `pixels[4096 * 3]`

**Comment by flok99** [comment]
Argh indeed. Unfortunately this problem also happens in a program where I didn't make this typo.
(044e31b979864578edaa185ce9e88a6718d6dc59 of https://github.com/flok99/constatus )
But with the *3 it doesn't happen in this smaller example.
I'll go and try to reproduce it with an other example.

**Comment by flok99** [comment]
==1818491== Possible data race during read of size 4 at 0x493340C by thread #115
==1818491== Locks held: none
==1818491==    at 0x48F20E4: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48A59B0: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48A7AFC: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48999E0: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48C9AEC: tjCompress2 (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x1095B6: use (thread_local.cpp:28)
==1818491==    by 0x1095B6: operator() (thread_local.cpp:42)
==1818491==    by 0x1095B6: __invoke_impl<void, main(int, char**)::<lambda()> > (invoke.h:60)
==1818491==    by 0x1095B6: __invoke<main(int, char**)::<lambda()> > (invoke.h:95)
==1818491==    by 0x1095B6: _M_invoke<0> (thread:244)
==1818491==    by 0x1095B6: operator() (thread:253)
==1818491==    by 0x1095B6: std::thread::_State_impl<std::thread::_Invoker<std::tuple<main::{lambda()#1}> > >::_M_run() (thread:196)
==1818491==    by 0x4A0ACB3: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==1818491==    by 0x4842B1A: ??? (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_helgrind-amd64-linux.so)
==1818491==    by 0x4B39608: start_thread (pthread_create.c:477)
==1818491==    by 0x4C75102: clone (clone.S:95)
==1818491== 
==1818491== This conflicts with a previous write of size 4 by thread #2
==1818491== Locks held: none
==1818491==    at 0x48F138D: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48F1498: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x489BA5E: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48A62A7: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48999D4: ??? (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x48C9AEC: tjCompress2 (in /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0)
==1818491==    by 0x1095B6: use (thread_local.cpp:28)
==1818491==    by 0x1095B6: operator() (thread_local.cpp:42)
==1818491==    by 0x1095B6: __invoke_impl<void, main(int, char**)::<lambda()> > (invoke.h:60)
==1818491==    by 0x1095B6: __invoke<main(int, char**)::<lambda()> > (invoke.h:95)
==1818491==    by 0x1095B6: _M_invoke<0> (thread:244)
==1818491==    by 0x1095B6: operator() (thread:253)
==1818491==    by 0x1095B6: std::thread::_State_impl<std::thread::_Invoker<std::tuple<main::{lambda()#1}> > >::_M_run() (thread:196)
==1818491==    by 0x4A0ACB3: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==1818491==  Address 0x493340c is in the Data segment of /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.2.0

Also in my example free(temp) is missing.

**Comment by flok99** [comment]
Apart from that ^ I cannot reproduced it. So indeed leave it closed.

**Comment by dcommander** [comment]
If you can use a debug build of the TurboJPEG API library and identify where in the code the possible data race is occurring, I can investigate that.  It may be an issue that was already fixed in 2.0.5.
